core/templates/pages/exploration-editor-page/services/exploration-rights.service.ts

```diff
@@ -247,6 +255,20 @@ export class ExplorationRightsService {
           response.rights.community_owned,
           response.rights.viewable_if_private
         );
+        this.alertsService.addSuccessMessage(
+          'Successfully removed user ' + memberUsername
+        );
+      })
+      .catch(error => {
+        categories.forEach(category => {
```

Comment:
You probably don't need to store the entire initialState, just store the category for which the user was removed, and you can then reference that again here.

Better still -- modify this function so that the category is passed into it (in addition to the username). That way, I think you'll have much simpler logic in this function.

Reply:
So one approach I'm trying is define a userRole variable, then set that equal to category once the memberUsername is removed:

 const categories = ['ownerNames', 'editorNames', 'viewerNames'] as const;
    let initialUsernamesByCategory: string[];
    let userRole = '';
    categories.forEach(category => {
      if (this[category]?.includes(memberUsername)) {
        initialUsernamesByCategory = [...this[category]];
        this[category] = this[category].filter(name => name !== memberUsername);
        userRole = category;
      }
    });
Then in the catch block, instead of having the for loop, its just one line: 
      this[userRole] = initialUsernamesByCategory. 
But I'm getting compile errors due to this[userRole], specifically:
 
core/templates/pages/exploration-editor-page/services/exploration-rights.service.ts(265,9): error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'ExplorationRightsService'.

  No index signature with a parameter of type 'string' was found on type 'ExplorationRightsService'.

I've tried defining index signatures in my previous commits because I've ran into this error before, but haven't had any luck in making it work properly and it also requires increasing the logic in other parts of the file.

The other approach that you mentioned of passing in the category to removeRoleAsync could work and would eliminate the need of having any for loops, however: 

removeRole(memberUsername: string, memberRole: string): void {
    this.alertsService.clearWarnings();

    const modalRef = this.ngbModal.open(RemoveRoleConfirmationModalComponent, {
      backdrop: true,
    });

    modalRef.componentInstance.username = memberUsername;
    modalRef.componentInstance.role = memberRole;

    modalRef.result.then(
      () => {
        this.explorationRightsService.removeRoleAsync(memberUsername);
        this.closeRolesForm();
      },
      () => {
        // Note to developers:
        // This callback is triggered when the Cancel button is
        // clicked. No further action is needed.
      }
    );
  }
  
  This is the function that calls removeRoleAsync. I noticed there is a memberRole defined, however these memberRoles map to hardcoded values in the corresponding html file. for example:
  (click)="removeRole(editorName, 'Collaborators')"
  
  I was thinking to map memberRole to a new category variable and then pass that into removeRoleAsync, (e.g. managers -> ownerNames, collaborators -> editorNames, playtesters -> viewerNames). However this would require some extra logic in removeRole to do this. 

I could also just modify the html file directly to change what the memberRole would be, or add a third parameter to removeRole that would be the specific category to pass in (editorNames, viewerNames, ownerNames).

However, these approaches will likely still cause the same compile error as above, specifically if I tried to do something like this:

  removeRoleAsync(memberUsername: string, memberRole: string): Promise<void> {
    let initialUsernamesByCategory: string[];
  initialUsernamesByCategory =  [...this[memberRole]];
  this[memberRole] = this.memberRole.filter(name => name !== memberUsername);
    });

this[memberRole] would throw that compile error.

Please let me know what you think about the different approaches, and also how I can fix the compile error. I can't use 'any' either as it fails the linter checks.

Reply:
@seanlip could you take a look at this and suggest which way would be best to move forward?

Reply:
Ok, I'm storing just the specific category, which is then referenced in the catch block.